engine/src/main/scala/cromwell/engine/backend/Backend.scala

```diff
@@ -108,9 +105,10 @@ trait Backend {
   def engineFunctions(fileSystems: List[FileSystem], workflowContext: WorkflowContext): WorkflowEngineFunctions
 
   /**
-   * Do any work that needs to be done <b>before</b> attempting to restart a workflow.
-   */
-  def prepareForRestart(restartableWorkflow: WorkflowDescriptor)(implicit ec: ExecutionContext): Future[Unit]
+    * Assume JobKey was in a 'Running' state when the server was shut down.
+    */
+  def isResumable(key: JobKey, executionInfo: Map[String, Option[String]]): Boolean = false
+  def isRestartable(key: JobKey, executionInfo: Map[String, Option[String]]): Boolean = false
```

Comment:
I'm not married to the way I implemented it, I ended up trying both ways and had an easier time working with this approach.  The way I thought about it is that you *can* return true from both of these.  However, the engine will always prefer to resume over restart.  Again, I'll change it if up really want.

The alternatives, as I see them:

```scala
def partitionRunningJobs(runningJobs: Set[ExecutionInfosByExecution]): (Set[Execution], Set[Execution])
```
or
```scala
def partitionRunningJobs(runningJobs: Map[Execution, Set[ExecutionInfo]]): (Set[Execution], Set[Execution])
```
or 
```scala
def partitionRunningJobs(runningJobs: Map[Execution, Map[String, Option[String]]): (Set[Execution], Set[Execution])
```
or
```scala
case class RestartableOrResumableExecutions(restartable: Set[Execution], resumable: Set[Execution])
def partitionRunningJobs(runningJobs: Set[ExecutionInfosByExecution]): RestartableOrResumableExecutions
```

... or any other combination of parameter(s) and return type.  Perhaps it's just a difference in personal taste, but I prefered the other scheme on mostly just aesthetics.  I'm not opposed to changing it, especially if @Horneth prefers one of the above approaches.

Another question that I didn't have a great answer for is whether or not to assume "if it's not resumable, then it's restartable".  If we take that assumption, then the whole function signature can be simplified.  So I just went with the more flexible model that you laid out in the ticket.

But given that we went with the backend deciding if something is 'restartable' or 'resumable', there was an issue of what to do with executions that fell into *neither* category of restartable or resumable.  Neither scheme handles it perfectly, but with this current scheme, I just [catch it in an else statement](https://github.com/broadinstitute/cromwell/pull/704/files#diff-b4e5dbe581881528f7565792c2766390R152) and fail the job.  With something more akin to partitioning, I guess I'd just have to check if the running execution is in neither `Set[Execution]`.